### PR TITLE
interfaces,overlord: support unversioned data

### PR DESCRIPTION
**Pull request by kyrofa**
The current versioned data directories are an important aspect of the upgrade/rollback strategy, but they don't scale well for applications that need to hold a vast amount of non-version-specific data (LP: [#1559248](https://bugs.launchpad.net/snappy/+bug/1559248)). This PR is an attempt to implement support for unversioned data (i.e. data that is shared between versions of a given snap), conforming to [Jamie's suggestion][2] (note that the naming is of course up for discussion). The shared directories are removed by overlord only once the final version is removed.

Note that this will also require changes to `ubuntu-core-launcher` to create the user-specific shared data directory (as it creates the versioned data directory). That can come once this PR is approved.

[2]: https://lists.ubuntu.com/archives/snappy-devel/2016-January/001451.html

### CODE REVIEW

File: overlord/snapstate/backend_test.go [hunk 1]
```diff
@@ -206,6 +206,14 @@ func (f *fakeSnappyBackend) RemoveSnapData(info *snap.Info) error {
 	return nil
 }
 
+func (f *fakeSnappyBackend) RemoveSnapSharedData(info *snap.Info) error {
+	f.ops = append(f.ops, fakeOp{
+		op:   "remove-snap-shared-data",
+		name: info.MountDir(),
```
Comment (zyga) [inline overlord/snapstate/backend_test.go]: MountDir feels odd here

File: snap/snapenv/snapenv.go [hunk 2]
```diff
@@ -63,6 +67,7 @@ func GetBasicSnapEnvVars(desc interface{}) []string {
 	return fillSnapEnvVars(desc, []string{
 		"SNAP={{.SnapPath}}",
 		"SNAP_DATA=/var{{.SnapPath}}",
+		"SNAP_SHARED_DATA={{CleanPath (printf \"/var%s/../shared\" .SnapPath)}}",
```
Comment (zyga) [inline snap/snapenv/snapenv.go]: Hmmm, I'm not sure this looks nice.

File: snap/snapenv/snapenv.go [hunk 3]
```diff
@@ -78,5 +83,6 @@ func GetBasicSnapEnvVars(desc interface{}) []string {
 func GetUserSnapEnvVars(desc interface{}) []string {
 	return fillSnapEnvVars(desc, []string{
 		"SNAP_USER_DATA={{.Home}}{{.SnapPath}}",
+		"SNAP_USER_SHARED_DATA={{CleanPath (printf \"%s%s/../shared\" .Home .SnapPath)}}",
```
Comment (zyga) [inline snap/snapenv/snapenv.go]: Ditto

File: snappy/snapdata.go [hunk 4]
```diff
@@ -45,6 +45,23 @@ func RemoveSnapData(snap *snap.Info) error {
...
+	dirs, err := snapSharedDataDirs(snap)
+	if err != nil {
+		return err
+	}
+
+	for _, dir := range dirs {
+		if err := os.RemoveAll(dir); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		os.Remove(filepath.Dir(dir))
```
Comment (zyga) [inline snappy/snapdata.go]: Based on
```
zyga@zyga-thinkpad-w510:~$ go doc os.RemoveAll
func RemoveAll(path string) error
    RemoveAll removes path and any children it contains. It removes everything
    it can but returns the first error it encounters. If the path does not
    exist, RemoveAll returns nil (no error).
```

I don't think you have to check for the `os.IsNotExist` or call `os.Remove` again later.

File: systemd/systemd_test.go [hunk 5]
```diff
@@ -211,8 +211,8 @@ X-Snappy=yes
 [Service]
 ExecStart=/usr/bin/ubuntu-core-launcher app aa-profile /apps/app/1.0/bin/start
 Restart=on-failure
-WorkingDirectory=/var/apps/app/1.0/
-Environment="SNAP=/apps/app/1.0/" "SNAP_DATA=/var/apps/app/1.0/" "SNAP_NAME=app" "SNAP_VERSION=1.0" "SNAP_REVISION=44" "SNAP_ARCH=%[3]s" "SNAP_LIBRARY_PATH=/var/lib/snapd/lib/gl:" "SNAP_USER_DATA=/root/apps/app/1.0/"
```
Comment (zyga) [inline systemd/systemd_test.go]: Wow, this stuff is stale! ```/apps/app/1.0``` Where did that come from?

In either case I don't think ```/var/appps``` is what we want today. I'd look at something in ```/var/lib/snapd/```

**Comment by zyga** [comment]
Some comments inline.

File: snap/snapenv/snapenv.go (hunk 2 above)
Comment (kyrofa) [inline snap/snapenv/snapenv.go]: Yeah I wasn't a huge fan either, but the only nice alternative I could think of was to stop being clever using `SnapPath` for everything and actually specify each of these explicitly, which is quite a bit more invasive (e.g. create an `info.BasePath()` to get `/snap/snapname` without the version, and so on). Would that be better?

File: overlord/snapstate/backend_test.go (hunk 1 above)
Comment (kyrofa) [inline overlord/snapstate/backend_test.go]: You mean because it includes the version? Perhaps this is reason enough to create an `info.BaseDir()` function, then.

File: snappy/snapdata.go (hunk 4 above)
Comment (kyrofa) [inline snappy/snapdata.go]: Agreed, this was taken from `RemoveSnapData()`. They should use the same function anyway, so I'll refactor them both.

File: systemd/systemd_test.go (hunk 5 above)
Comment (kyrofa) [inline systemd/systemd_test.go]: Indeed, lots of test cruft here. Note that it's using version instead of revision as well. I thought cleaning it would make for a noisy diff though. Should I anyway?

File: overlord/snapstate/backend_test.go (hunk 1 above)
Comment (zyga) [inline overlord/snapstate/backend_test.go]: My comment was based on just the name: ... MountDir connection but I see that other tests use this as well.

File: systemd/systemd_test.go (hunk 5 above)
Comment (zyga) [inline systemd/systemd_test.go]: Perhaps as a side branch. I think it's well worth not adding more /apps test data.

File: overlord/snapstate/backend_test.go (hunk 1 above)
Comment (kyrofa) [inline overlord/snapstate/backend_test.go]: Yeah, it's nice for testing that a given operation happens on a given version of the snap. We may want to leave this as `MountDir()`, then.

File: systemd/systemd_test.go (hunk 5 above)
Comment (kyrofa) [inline systemd/systemd_test.go]: Alright, I'll leave this alone here, then.

File: interfaces/apparmor/template.go [hunk 6]
```diff
@@ -233,13 +233,15 @@ var defaultTemplate = []byte(`
 
   # Writable home area for this version.
   owner @{HOME}/snap/@{SNAP_NAME}/@{SNAP_REVISION}/** wl,
+  owner @{HOME}/snap/@{SNAP_NAME}/shared/** wl,
```
Comment (niemeyer) [inline interfaces/apparmor/template.go]: Although I'd like to sync up with everyone on this, the approach looks nice and simple.

One initial comment: shared is perhaps not the best word. It also means "make it public" (share with whom?), and it may imply concurrent use which will never be the case here (only the active snap will touch it). As a preliminary strawman, I'd suggest "common".

File: snap/snapenv/snapenv.go (hunk 2 above)
Comment (kyrofa) [inline snap/snapenv/snapenv.go]: Fixed.

File: snap/snapenv/snapenv.go (hunk 3 above)
Comment (kyrofa) [inline snap/snapenv/snapenv.go]: Also fixed.

File: snappy/snapdata.go (hunk 4 above)
Comment (kyrofa) [inline snappy/snapdata.go]: Done. Though note: the later `os.Remove()` call is what removes the parent directories when they're empty, so I've left that, but added a comment to clarify.

File: interfaces/apparmor/template.go (hunk 6 above)
Comment (kyrofa) [inline interfaces/apparmor/template.go]: > shared is perhaps not the best word. It also means "make it public" (share with whom?), and it may imply concurrent use which will never be the case here (only the active snap will touch it).

Great point, I agree.

> As a preliminary strawman, I'd suggest "common".

I'm good with "common." How about the variable name format: `SNAP_COMMON_DATA` and `SNAP_USER_COMMON_DATA`? I'm concerned the latter gives the impression that the data is common among users, which of course isn't the case. Can you think of something more clear?

File: overlord/snapstate/backend_test.go (hunk 1 above)
Comment (niemeyer) [inline overlord/snapstate/backend_test.go]: Yeah, let's not overthink this too much for now. It's just a way to print out what the backend had at hand when the given method was called.

File: overlord/snapstate/snapmgr.go [hunk 7]
```diff
@@ -289,7 +289,18 @@ func (m *SnapManager) doClearSnapData(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
-	return m.backend.RemoveSnapData(info)
+	if err = m.backend.RemoveSnapData(info); err != nil {
+		return err
+	}
+
+	// Only remove data shared between versions if this is the last version
```
Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: This is a bit problematic, and makes me want to think further about the whole approach we're taking. Imagine a button that says "Reset snap data".. if we keep the shared data around we're not actually resetting the data. We'll restart the snap, and its data is still there, partially! How do we reset the data for a snap, ever, once we have a shared directory?

File: interfaces/apparmor/template.go (hunk 6 above)
Comment (kyrofa) [inline interfaces/apparmor/template.go]: I guess "unversioned" is an option as well.

File: snap/info.go [hunk 8]
```diff
@@ -36,7 +36,10 @@ type PlaceInfo interface {
 	// Name returns the name of the snap.
 	Name() string
 
-	// MountDir returns the base directory of the snap.
+	// BaseDir returns the base directory for all versions of the snap.
+	BaseDir() string
```
Comment (niemeyer) [inline snap/info.go]: We had a base dir very recently, which got renamed because it was unclear. It doesn't feel much more clear now. Which base dir?

File: overlord/snapstate/snapmgr.go (hunk 7 above)
Comment (kyrofa) [inline overlord/snapstate/snapmgr.go]: That sounds like new functionality, though-- do we currently support clearing data out from under an installed snap? As part of that feature I expected we'd also clear the shared data.

File: snap/info.go [hunk 9]
```diff
@@ -45,18 +48,29 @@ type PlaceInfo interface {
...
+	SharedDataHomeDir() string
 }
 
 // MinimalPlaceInfo returns a PlaceInfo with just the location information for a snap of the given name and revision.
 func MinimalPlaceInfo(name string, revision int) PlaceInfo {
 	return &Info{SideInfo: SideInfo{OfficialName: name, Revision: revision}}
 }
 
+// BaseDir returns the base directory for all versions of the snap.
+func BaseDir(name string) string {
```
Comment (niemeyer) [inline snap/info.go]: Same as prior point above.

File: snap/info.go [hunk 10]
```diff
@@ -142,6 +156,11 @@ func (s *Info) strRevno() string {
 	return strconv.Itoa(s.Revision)
 }
 
+// BaseDir returns the base directory for all versions of the snap.
+func (s *Info) BaseDir() string {
```
Comment (niemeyer) [inline snap/info.go]: Same as prior point above.

File: snap/info.go (hunk 8 above)
Comment (kyrofa) [inline snap/info.go]: Agreed. This is part of why I used filepath.Clean initially-- this directory really has no meaning except for building paths. `ParentDir()`? `MountsDir()`? `MountParentDir()`?

File: snap/snapenv/snapenv.go [hunk 11]
```diff
@@ -63,6 +63,7 @@ func GetBasicSnapEnvVars(desc interface{}) []string {
 	return fillSnapEnvVars(desc, []string{
 		"SNAP={{.SnapPath}}",
 		"SNAP_DATA=/var{{.SnapPath}}",
+		"SNAP_SHARED_DATA=/var{{.SnapBasePath}}/shared",
```
Comment (niemeyer) [inline snap/snapenv/snapenv.go]: Why are we building this variable here when we have a method in info that returns the exact correct path? We should have the info here and use its methods (templates can do that) instead of inventing a new way to get the same data for each one of these variables.

Comment (kyrofa) [inline snap/snapenv/snapenv.go]: That would involve adding `stripGlobalRootDir` as a template function which makes this a little less pretty. Is that okay?

Comment (kyrofa) [inline snap/snapenv/snapenv.go]: Note that touches a lot of tests as well. It may be better suited for another PR.

File: snappy/binaries.go [hunk 12]
```diff
@@ -71,29 +71,31 @@ ubuntu-core-launcher {{.UdevAppName}} {{.AaProfile}} {{.Target}} "$@"
...
-		AppName:     app.Name,
-		SnapArch:    arch.UbuntuArchitecture(),
-		SnapPath:    pkgPath,
-		Version:     app.Snap.Version,
-		Revision:    app.Snap.Revision,
-		UdevAppName: app.SecurityTag(),
-		Home:        "$HOME",
-		Target:      actualBinPath,
-		AaProfile:   app.SecurityTag(),
+		SnapName:     app.Snap.Name(),
```
Comment (niemeyer) [inline snappy/binaries.go]: Let's please have a single field here, "App", and kill all of these fields which are simply accessing it by doing so in the template itself.

File: snappy/binaries.go [hunk 13]
```diff
@@ -71,29 +71,31 @@ ubuntu-core-launcher {{.UdevAppName}} {{.AaProfile}} {{.Target}} "$@"
...
+		AppName:      app.Name,
+		SnapArch:     arch.UbuntuArchitecture(),
+		SnapBasePath: stripGlobalRootDir(app.Snap.BaseDir()),
+		SnapPath:     pkgPath,
+		Version:      app.Snap.Version,
+		Revision:     app.Snap.Revision,
+		UdevAppName:  app.SecurityTag(),
+		Home:         "$HOME",
```
Comment (niemeyer) [inline snappy/binaries.go]: This also shouldn't be here. We can have the string in the template itself.

Comment (kyrofa) [inline snappy/binaries.go]: Well, it changes to `/root` if it's a service, though.

File: snappy/binaries.go (hunk 12 above)
Comment (kyrofa) [inline snappy/binaries.go]: +1, but ditto my comment about `stripGlobalRootDir` and the fact that this will add a lot of diff noise. Do you want me to do that here, or in another PR?

File: overlord/snapstate/snapmgr.go (hunk 7 above)
Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: Okay, the answer is that we'll need a separate task that kills the data for all snaps at once. If we intentionally want to reset the data for a snap, it should do so for all snaps rather than just the current revision. The experience would probably be awkward otherwise.

So this is fine for now, thanks.

Comment (kyrofa) [inline overlord/snapstate/snapmgr.go]: Agreed.

File: interfaces/apparmor/template.go (hunk 6 above)
Comment (niemeyer) [inline interfaces/apparmor/template.go]: Quickly discussed this with the team and the response was positive.

So let's go ahead with the proposal, just replacing the "shared" term by "common" everywhere please.

File: snap/info.go (hunk 8 above)
Comment (niemeyer) [inline snap/info.go]: Let's please drop all BaseDir entries. The use case in the template is better handled by exposing App to the template directly, and using the proper method names instead of rebuilding the paths inside the template.

File: snap/snapenv/snapenv.go (hunk 11 above)
Comment (niemeyer) [inline snap/snapenv/snapenv.go]: Per prior note, let's expose the App to the template and use the existing methods directly here instead of rebuilding the path locally.

Comment (kyrofa) [inline snap/snapenv/snapenv.go]: Sounds good.

**Comment by kyrofa** [comment]
As agreed in Telegram, I've pushed updates renaming "shared" to "common," and removed the variables from the launchers.

File: snappy/binaries.go (hunk 12 above)
Comment (zyga) [inline snappy/binaries.go]: I can start a parallel branch that does that and we can see where it leads.

File: snappy/binaries.go (hunk 13 above)
Comment (zyga) [inline snappy/binaries.go]: That is still a ``"$HOME"`` constant in the template.

Comment (kyrofa) [inline snappy/binaries.go]: I'm talking about https://github.com/ubuntu-core/snappy/blob/master/systemd/systemd.go#L388 . You can't use $HOME or %h in the service file.

**Comment by niemeyer** [comment]
LGTM

File: interfaces/apparmor/template.go [hunk 14]
```diff
@@ -233,13 +233,15 @@ var defaultTemplate = []byte(`
...
   owner @{HOME}/snap/@{SNAP_NAME}/@{SNAP_REVISION}/** wl,
+  owner @{HOME}/snap/@{SNAP_NAME}/common/** wl,
 
   # Read-only system area for other versions
   /var/snap/@{SNAP_NAME}/   r,
   /var/snap/@{SNAP_NAME}/** mrkix,
 
   # Writable system area only for this version
   /var/snap/@{SNAP_NAME}/@{SNAP_REVISION}/** wl,
+  /var/snap/@{SNAP_NAME}/common/** wl,
```
Comment (jdstrand) [inline interfaces/apparmor/template.go]: FYI, these changes look fine security policy-wise.

File: snap/info.go [hunk 15]
```diff
@@ -45,8 +45,14 @@ type PlaceInfo interface {
 	// DataDir returns the data directory of the snap.
 	DataDir() string
 
+	// CommonDataDir returns the data directory common between versions of the snap.
```
Comment (zyga) [inline snap/info.go]: s/versions/revisions/

File: snap/info.go [hunk 16]
```diff
@@ -45,8 +45,14 @@ type PlaceInfo interface {
...
+	CommonDataDir() string
+
 	// DataHomeDir returns the per user data directory of the snap.
 	DataHomeDir() string
+
+	// CommonDataHomeDir returns the per user data directory common between versions of the snap.
```
Comment (zyga) [inline snap/info.go]: ditto (revisions)

File: snap/info.go [hunk 17]
```diff
@@ -157,11 +163,21 @@ func (s *Info) DataDir() string {
 	return filepath.Join(dirs.SnapDataDir, s.Name(), s.strRevno())
 }
 
+// CommonDataDir returns the data directory common between versions of the snap.
```
Comment (zyga) [inline snap/info.go]: revisions

File: snap/info.go [hunk 18]
```diff
@@ -157,11 +163,21 @@ func (s *Info) DataDir() string {
...
+	return filepath.Join(dirs.SnapDataDir, s.Name(), "common")
+}
+
 // DataHomeDir returns the per user data directory of the snap.
 func (s *Info) DataHomeDir() string {
 	return filepath.Join(dirs.SnapDataHomeGlob, s.Name(), s.strRevno())
 }
 
+// CommonDataHomeDir returns the per user data directory common between versions of the snap.
+func (s *Info) CommonDataHomeDir() string {
```
Comment (zyga) [inline snap/info.go]: revisions

File: snap/info.go (hunk 15 above)
Comment (niemeyer) [inline snap/info.go]: Good catch. Even better: "common across revisions"

File: snap/info.go [hunk 19]
```diff
@@ -157,11 +163,21 @@ func (s *Info) DataDir() string {
...
+	return filepath.Join(dirs.SnapDataHomeGlob, s.Name(), "common")
```
Comment (zyga) [inline snap/info.go]: This would return ``/home/*/snap/$snap/common``. This is not a new problem but it's not clear that the returned value is a glob.

File: snap/info.go [hunk 20]
```diff
@@ -157,11 +163,21 @@ func (s *Info) DataDir() string {
...
 }
 
+// CommonDataDir returns the data directory common between versions of the snap.
+func (s *Info) CommonDataDir() string {
+	return filepath.Join(dirs.SnapDataDir, s.Name(), "common")
+}
+
 // DataHomeDir returns the per user data directory of the snap.
 func (s *Info) DataHomeDir() string {
 	return filepath.Join(dirs.SnapDataHomeGlob, s.Name(), s.strRevno())
```
Comment (zyga) [inline snap/info.go]: FYI: note the glob remark on `CommonDataHomeDir` below

File: snap/info.go (hunk 19 above)
Comment (kyrofa) [inline snap/info.go]: Indeed. I think both this and `DataHomeDir()` should have "Glob" at the end of their names. I can do that here if you like.

File: snap/info.go (hunk 15 above)
Comment (kyrofa) [inline snap/info.go]: Done.

File: snap/info.go (hunk 16 above)
Comment (kyrofa) [inline snap/info.go]: Done.

File: snap/info.go (hunk 17 above)
Comment (kyrofa) [inline snap/info.go]: Done.

File: snap/info.go (hunk 18 above)
Comment (kyrofa) [inline snap/info.go]: Done.

File: snap/info.go (hunk 19 above)
Comment (niemeyer) [inline snap/info.go]: Instead of having a Glob in the name, let's please make it take a username parameter instead which can be set to "*" in current use cases.

Comment (kyrofa) [inline snap/info.go]: I actually did that once as well. You run into problems when you try to use '$HOME' with a global root.

Comment (kyrofa) [inline snap/info.go]: Also, you actually need the globs for removal, so that would require new functions.

**Comment by zyga** [comment]
LGTM

File: snap/info.go (hunk 19 above)
Comment (pedronis) [inline snap/info.go]: DataHomeDir is my fault refactoring away and not noticing that it involved actually a glob, not sure one method and one dirs pkg variable can cover all our use cases though, it seems we need to produce:
<globRootDir>/home/*/snap/...   "/root/snap/..."  and "$HOME/snap/..."

Comment (niemeyer) [inline snap/info.go]: We already have that problem today, so let's merge this and tackle that separately.

File: interfaces/apparmor/template.go (hunk 6 above)
Comment (didrocks) [inline interfaces/apparmor/template.go]: Reading about this, "common" makes sense from an internal code perspective for sure, I'm less sure from a developer perspective it would though.
"common" is implicit (as for "shared") that multiple entities can access to it at the same time, which isn't the case as @niemeyer suggested.

What you do actually end up with are versioned and version-independant areas. It's even more obvious when you ls in the data directory as you will have:
`version1/`, `version2/`, …, `current`. Why not then rather using `unversioned` which I guess will live better next to those versions and the current symlink? Thoughts?

Comment (niemeyer) [inline interfaces/apparmor/template.go]: "common" doesn't really have any access-related concurrency connotation that I'm aware of, and searching for these terms seems to agree with that. Do you have any reference pointing out otherwise?

The fact multiple entities can access it is intended, though. That's exactly what "not versioned" means, and it's always more pleasing to label something after what it means than after what it does not mean.

Comment (didrocks) [inline interfaces/apparmor/template.go]: I partially agree with it's more pleasing to label something in a non negative way. However, I do fell that "unversioned" underly a fact that "common" doesn't hilight well enough: you need to ensure data you put there are  forward compatible.

The "unversioned" term will be a little bit more scary to developers for this and it's a good reason: you really need to ensure that data you put there are working with any available version of your application. You don't want non forward-compatible schema of your database for instance (I'm afraid that this directory will be used a lot for databases in particular).
"common" is a little bit more gentle in this regard and don't underline this risk… at least from a non native speaker perspective (so take that with a grain of salt ;))
